src/modules/dashboard/components/CoinMachine/RemainingDisplayWidgets/RemainingTokens.tsx

```diff
@@ -0,0 +1,137 @@
+import React, { useMemo } from 'react';
+import { FormattedMessage, defineMessages } from 'react-intl';
+import { BigNumber } from 'ethers/utils';
+
+import { getFormattedTokenValue } from '~utils/tokens';
+
+import { TokenPriceStatuses } from '../TokenPriceStatusIcon/TokenPriceStatusIcon';
+
+import RemainingWidget from './RemainingDisplayWidget';
+
+export enum DataDisplayType {
+  Time = 'Time',
+  Tokens = ' Tokens',
+}
+
+type Appearance = {
+  theme?: 'white' | 'danger';
+};
+
+type Props = {
+  appearance?: Appearance;
+  periodTokens?: {
+    decimals: number;
+    soldPeriodTokens: BigNumber;
+    maxPeriodTokens: BigNumber;
+    targetPeriodTokens: BigNumber;
+  };
+};
+
+const displayName =
+  'dashboard.CoinMachine.RemainingDisplayWidgets.RemainingTokens';
+
+const MSG = defineMessages({
+  tokensRemainingTitle: {
+    id: 'dashboard.CoinMachine.RemainingDisplayWidgets.RemainingTokens.title',
+    defaultMessage: 'Tokens remaining',
+  },
+  tokensRemainingTooltip: {
+    id: 'dashboard.CoinMachine.RemainingDisplayWidgets.RemainingTokens.tooltip',
+    defaultMessage: `This is the number of tokens remaining in the current batch.`,
+  },
+  tokensTypePlaceholder: {
+    id: 'dashboard.CoinMachine.RemainingDisplayWidgets.RemainingTokens.title',
+    defaultMessage: '0',
+  },
+  tokensTypeFooterText: {
+    id: `dashboard.CoinMachine.RemainingDisplayWidgets.RemainingTokens.footerText`,
+    defaultMessage: 'Price next sale',
+  },
+  soldOut: {
+    id: 'dashboard.CoinMachine.RemainingDisplayWidgets.RemainingTokens.soldOut',
+    defaultMessage: 'SOLD OUT',
+  },
+});
+
+const RemainingTokens = ({
+  appearance = { theme: 'white' },
+  periodTokens,
+}: Props) => {
+  const widgetText = useMemo(() => {
+    return {
+      title: MSG.tokensRemainingTitle,
+      placeholder: MSG.tokensTypePlaceholder,
+      tooltipText: MSG.tokensRemainingTooltip,
+      footerText: MSG.tokensTypeFooterText,
+    };
+  }, []);
+
+  const displayedValue = useMemo(() => {
+    if (periodTokens) {
+      const { soldPeriodTokens, decimals, maxPeriodTokens } = periodTokens;
```

Comment:
@rdig, this will be fixed when Armando merges his PR about the `isCurrentlyOnSale` logic and I rebase on it. There is a check that if it's not on sale then the tokens return undefined (line 94 in `CoinMachine`) and in `RemainingTokens` if the periodTokens are undefined, then `0` will be shown.

After I rebase and before I merge I will make sure that this logic works. For now you can manually check it by changing `isSale` variable in `CoinMachine` to false.

Reply:
Ok then